tests/test_enum.rs

```diff
@@ -63,3 +71,93 @@ fn test_default_repr_correct() {
         py_assert!(py, var2, "repr(var2) == 'MyEnum.OtherVariant'");
     })
 }
+
+#[pyclass]
+enum CustomDiscriminant {
+    One = 1,
+    Two = 2,
+}
+
+#[test]
+fn test_custom_discriminant() {
+    Python::with_gil(|py| {
+        #[allow(non_snake_case)]
+        let CustomDiscriminant = py.get_type::<CustomDiscriminant>();
+        let one = Py::new(py, CustomDiscriminant::One).unwrap();
+        let two = Py::new(py, CustomDiscriminant::Two).unwrap();
+        py_run!(py, CustomDiscriminant one two, r#"
+        assert CustomDiscriminant.One == one
+        assert CustomDiscriminant.Two == two
+        assert one != two
+        "#);
+    })
+}
+
+#[test]
+fn test_enum_to_int() {
+    Python::with_gil(|py| {
+        let one = Py::new(py, CustomDiscriminant::One).unwrap();
+        py_assert!(py, one, "int(one) == 1");
+        let v = Py::new(py, MyEnum::Variant).unwrap();
+        let v_value = MyEnum::Variant as isize;
+        py_run!(py, v v_value, "int(v) == v_value");
+    })
+}
+
+#[test]
+fn test_enum_compare_int() {
+    Python::with_gil(|py| {
+        let one = Py::new(py, CustomDiscriminant::One).unwrap();
+        py_run!(
+            py,
+            one,
+            r#"
+            assert one == 1
+            assert 1 == one
+            assert one != 2
+        "#
+        )
+    })
+}
+
+#[pyclass]
+#[repr(u8)]
+enum SmallEnum {
+    V = 1,
+}
+
+#[test]
+fn test_enum_compare_int_no_throw_when_overflow() {
+    Python::with_gil(|py| {
+        let v = Py::new(py, SmallEnum::V).unwrap();
+        py_assert!(py, v, "v != 1<<30")
+    })
+}
+
+#[pyclass]
+#[repr(usize)]
+#[allow(clippy::enum_clike_unportable_variant)]
+enum BigEnum {
+    V = usize::MAX,
+}
+
+#[test]
+fn test_big_enum_no_overflow() {
+    Python::with_gil(|py| {
+        let usize_max = usize::MAX;
+        let v = Py::new(py, BigEnum::V).unwrap();
+        py_assert!(py, usize_max v, "v == usize_max");
+        py_assert!(py, usize_max v, "int(v) == usize_max");
+    })
+}
+
+#[pyclass]
+#[repr(u16, align(8))]
+enum TestReprParse {
+    V,
+}
+
+#[test]
+fn test_repr_parse() {
+    assert_eq!(std::mem::align_of::<TestReprParse>(), 8);
+}
```

Comment:
As a final thought, it would be awesome to add some UI tests to `tests/ui/invalid_pyclass_enum.rs` to check the error messages when using enums with data which fall in the #417 category.

(I know you plan to work on #417 soon, though we might have the 0.16 release in the middle and it'd be great to know we have a good UX for these in the interim.)